agent/system_metrics.go

```diff
@@ -0,0 +1,215 @@
+package agent
+
+import (
+	"context"
+	"runtime"
+	"sync"
+	"time"
+
+	"bytes"
+	"encoding/json"
+
+	"github.com/evergreen-ci/evergreen/rest/client"
+	system_metrics "github.com/evergreen-ci/timber/system_metrics"
+	"github.com/mongodb/ftdc/metrics"
+	"github.com/pkg/errors"
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/shirou/gopsutil/disk"
+	"github.com/shirou/gopsutil/host"
+	"github.com/shirou/gopsutil/process"
+)
+
+// MetricCollector is an interface representing an object that can collect
+// a single system metric over a series of time steps.
+type MetricCollector interface {
+	// Name returns a string indicating the type of metric collected, such as "uptime".
+	Name() string
+	// Format returns the format of the collected data.
+	Format() DataFormat
+	// Collect returns the value of the collected metric when the function is called.
+	Collect() ([]byte, error)
+}
+
+type SystemMetricCollector struct {
+	mu          sync.Mutex
+	wg          sync.WaitGroup
+	innerCancel context.CancelFunc
+	logger      client.LoggerProducer
+	taskOpts    *system_metrics.SystemMetricsOptions
+	connOpts    system_metrics.ConnectionOptions
+	interval    time.Duration
+	collectors  []MetricCollector
+	id          string
+	client      *system_metrics.SystemMetricsClient
+	errored     bool
+	closed      bool
+}
+
+// DataFormat describes how the time series data is stored.
+type DataFormat int32
+
+// Valid DataFormat values.
+const (
+	DataFormatText DataFormat = 0
+	DataFormatFTDC DataFormat = 1
+	DataFormatBSON DataFormat = 2
+	DataFormatJSON DataFormat = 3
+	DataFormatCSV  DataFormat = 4
+)
+
+type diskUsageCollector struct{}
+
+func NewDiskUsageCollector() *diskUsageCollector {
+	return new(diskUsageCollector)
+}
+
+func (collector *diskUsageCollector) Name() string { return "disk_usage" }
+
+func (collector *diskUsageCollector) Format() DataFormat { return DataFormatFTDC }
+
+func (collector *diskUsageCollector) Collect(ctx context.Context, dir string) ([]byte, error) {
+	usage, err := disk.Usage(dir)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem capturing metrics with gopsutil")
+	}
+
+	return convertJSONToFTDC(ctx, usage)
+}
+
+func convertJSONToFTDC(ctx context.Context, metric interface{}) ([]byte, error) {
+	jsonMetrics, err := json.Marshal(metric)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem converting metrics to JSON")
+	}
+
+	opts := metrics.CollectJSONOptions{
+		InputSource:   bytes.NewReader(jsonMetrics),
+		SampleCount:   1,
+		FlushInterval: 1 * time.Second,
+	}
+
+	output, err := metrics.CollectJSONStream(ctx, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem converting FTDC to JSON")
+	}
+	return output, nil
+}
+
+type uptimeCollector struct{}
+
+type UptimeWrapper struct {
```

Comment:
Unless we have a reason to export this, we shouldn't export it since it's not used outside this package.

Reply:
done